### ISSUE TITLE
Highlights Flickering

### ISSUE THREAD

**Issue by tiagovla** [theHamsta/nvim-semantic-tokens#7]
Should ``on_invalidate_range/clear_highlights`` be called on every event?

![gifcast_220625152527](https://user-images.githubusercontent.com/30515389/175786619-5bde669d-d651-4782-ab60-d06a15f23ab6.gif)

**Comment by theHamsta** [comment]
@tiagovla it is only called when the language server tells me to refresh. Can you check whether the LSP actually calls this method https://github.com/theHamsta/neovim/blob/5099653ae1089e9180c560b7a4172674d0fa5398/runtime/lua/vim/lsp/handlers.lua#L25-L27 (e.g. by removing the handler). Maybe we should consider to not set this handler. It does not flicker for the language server I use (clang, sumneko_lua, rust-analyzer).

**Comment by theHamsta** [comment]
@[jdrouhard](https://github.com/jdrouhard) did you ever experience something similar?

**Comment by tiagovla** [comment]
> @tiagovla it is only called when the language server tells me to refresh. Can you check whether the LSP actually calls this method https://github.com/theHamsta/neovim/blob/5099653ae1089e9180c560b7a4172674d0fa5398/runtime/lua/vim/lsp/handlers.lua#L25-L27 (e.g. by removing the handler). Maybe we should consider to not set this handler. It does not flicker for the language server I use (clang, sumneko_lua, rust-analyzer).
>
I checked and it's not the ``on_refresh``.

> 
> What is the language server you're using?
> 
Pylance

> EDIT: I see it calls `on_invalidate_range/clear_highlights` on cursor move. Will check the upstream PR later. You should maybe link this issue upstream that `on_invalidate_range/clear_highlights` is called to often. Actually it should only clear when it already has the highlight so the flickering is strange...
>

It is indeed the ``on_invalidate_range/clear_highlights``. Clangd works fine for me (it doesn't call it on every event).

**Comment by tiagovla** [comment]
I was counting the number of times ``on_invalidate_range`` was called using clangd.
![gifcast_220625172545](https://user-images.githubusercontent.com/30515389/175789606-972388ea-4352-43c3-ad6d-3a17cbe09f40.gif)

**Comment by badicsalex** [comment]
I encountered a similar thing with rust_analyzer. If I add or remove multiple lines in quick succession, there are multiple in-flight token requests. The tick check fails for the first response, which leaves the buffer un-highlighted momentarily, but then the 'correct' response arrives, and highlighting is correct again.

For now all I did was move the invalidate call below the error checking if, which seems to solve my problem. "Leftover" highlights are not really a problem IMO, but a stricter way to do this would be to only call invalidate if there's an actual error, since with the tick check it can be assumed that the response with the correct tick is on the way.
